lib/wasi/src/state/types.rs

```diff
@@ -970,6 +970,77 @@ impl WasiFile for Stdin {
     }
 }
 
+/// For piping stdio. Stores all output / input in a byte-vector.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Pipe {
+    buffer: Vec<u8>,
```

Comment:
It probably makes sense to use a [`std::collections::VecDeque`](https://doc.rust-lang.org/std/collections/struct.VecDeque.html) here so we can pop off the front and append to the back in constant time